### PR TITLE
Functionality to remove batch NAs from PCs

### PR DESCRIPTION
**Pull request by cbethell**
Per https://github.com/AlexsLemonade/scpcaTools/pull/181#discussion_r1262956522, this PR adds a function to `remove_batch_nas_from_pcs` and implements this function in the `calculate_silhouette_width()` function added in PR #181.

### CODE REVIEW

File: R/remove_batch_nas_from_pcs.R [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  # Set PC rownames to be the batches
+  rownames(integrated_pcs) <- batches
+
+  return(
+    list(
+      pcs = integrated_pcs,
+      indices = retain_indices
+    )
```
Comment (allyhawkins) [inline R/remove_batch_nas_from_pcs.R:27]: I think we should just return the pcs. I don't think we need the indices because we don't actually use those?

**Review by allyhawkins — COMMENTED** [review]
I have some thoughts on naming and purpose. I think we should slightly reframe this to focus on setting the names of the pcs. Not a lot of code changes, but that should be reflected in the function's documentation.

File: R/remove_batch_nas_from_pcs.R [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#' Remove NA batch cells from PCs for use in integration metric calculations
+#'
+#' @param integrated_pcs The PCs with NA batch cells to be removed
+#' @param batches Vector of cell-wise batch information whose length is the number of
+#'   rows in `integrated_pcs`
+#'
+#' @return List with two items: `pcs`: PCs with NA batch cells removed and
+#'   rownames assigned as batch; `indices`: Which rows were _retained_
+remove_batch_nas_from_pcs <- function(integrated_pcs, batches) {
```
[the remark this answers is not in the document]
Comment (cbethell) [reply R/remove_batch_nas_from_pcs.R:9]: Yes I like this idea!

File: R/remove_batch_nas_from_pcs.R (hunk 1 above)
Comment (cbethell) [reply R/remove_batch_nas_from_pcs.R]: Okay, I made this modification -- let me know if I modified things as you would have expected @allyhawkins

File: R/calculate_silhouette_width.R [hunk 3]
```diff
@@ -45,11 +45,13 @@ calculate_silhouette_width <- function(integrated_sce,
   # Pull out the PCs or analogous reduction
   pcs <- reducedDim(integrated_sce, pc_name)
 
+  # Label rownames and remove batch NAs from PCs
+  labeled_pcs <- set_pc_rownames(pcs, colData(integrated_sce)[,batch_column])[["pcs"]]
```
Comment (allyhawkins) [inline R/calculate_silhouette_width.R:49]: Before this we should check that the `batch_column` is in the `colData`.

File: R/set_pc_rownames.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#' Label rownames of provided PCs and remove NA batch cells from PCs
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:1]:
```suggestion
#' Label rownames of provided PCs and remove cells with NA batch label from PCs
```

File: R/set_pc_rownames.R [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+#'
+#' @param integrated_pcs The PCs with unlabeled rownames
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:3]:
```suggestion
#' @param pcs The PCs with unlabeled rownames
```
This doesn't have to be integrated.

File: R/set_pc_rownames.R [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#' @param batches Vector of cell-wise batch information whose length is the number of
+#'   rows in `integrated_pcs`
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:5]:
```suggestion
#'   rows in `pcs`
```

File: R/set_pc_rownames.R [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#'
+#' @return PCs with labeled rownames and NA batch cells removed
+set_pc_rownames <- function(integrated_pcs, batches) {
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:8]:
```suggestion
set_pc_rownames <- function(pcs, batches) {
```

File: R/set_pc_rownames.R [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  retain_indices <- which(!is.na(batches))
+  batches <- batches[retain_indices]
+  integrated_pcs <- integrated_pcs[retain_indices,]
+
+  # Check dimensions
+  if (nrow(integrated_pcs) != length(batches)) {
+    stop("Incompatable PC and batch information dimensions after removing NAs.")
+  }
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:17]: I feel like this should be before the `retain_indices` part.

File: R/set_pc_rownames.R [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+#' @param integrated_pcs The PCs with unlabeled rownames
+#' @param batches Vector of cell-wise batch information whose length is the number of
+#'   rows in `integrated_pcs`
+#'
+#' @return PCs with labeled rownames and NA batch cells removed
+set_pc_rownames <- function(integrated_pcs, batches) {
+
+  retain_indices <- which(!is.na(batches))
+  batches <- batches[retain_indices]
+  integrated_pcs <- integrated_pcs[retain_indices,]
```
Comment (allyhawkins) [inline R/set_pc_rownames.R:12]:
```suggestion
  pcs <- pcs[retain_indices,]
```

**Review by allyhawkins — COMMENTED** [review]
Just a few smaller comments to address, but the reorganization looks good.

**Comment by cbethell** [comment]
Awesome, implemented those last set of comments and updates the docs! 👍

**Review by allyhawkins — APPROVED** [review]
Looks good 👍

**Pull request merged**